internal/participation/service.go

```diff
@@ -0,0 +1,71 @@
+package participation
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/CollActionteam/collaction_backend/internal/constants"
+	m "github.com/CollActionteam/collaction_backend/internal/models"
+	"github.com/CollActionteam/collaction_backend/models"
+	"github.com/CollActionteam/collaction_backend/pkg/repository"
+	"github.com/CollActionteam/collaction_backend/utils"
+)
+
+type Service interface {
+	GetParticipation(ctx context.Context, userID string, crowdactionID string) (*m.ParticipationRecord, error)
+	RegisterParticipation(ctx context.Context, userID string, name string, crowdaction *models.Crowdaction, payload m.JoinPayload) error
+	CancelParticipation(ctx context.Context, userID string, crowdaction *models.Crowdaction) error
+}
+
+type participationService struct {
+	participationRepository repository.Participation
+}
+
+func NewParticipationService(participationRepository repository.Participation) Service {
+	return &participationService{
+		participationRepository: participationRepository,
+	}
+}
+
+func recordEvent(userID string, crowdactionID string, commitments []string, count int) error {
+	qc := utils.CreateQueueClient()
+	event := m.ParticipationEvent{
+		UserID:        userID,
+		CrowdactionID: crowdactionID,
+		Commitments:   commitments,
+		Count:         count,
+	}
+	return utils.SendQueueMessage(qc, constants.ParticipationQueueName, event)
+}
+
+func (s *participationService) GetParticipation(ctx context.Context, userID string, crowdactionID string) (*m.ParticipationRecord, error) {
+	return s.participationRepository.Get(ctx, userID, crowdactionID)
+}
+
+func (s *participationService) RegisterParticipation(ctx context.Context, userID string, name string, crowdaction *models.Crowdaction, payload m.JoinPayload) error {
+	if !utils.IsFutureDateString(crowdaction.DateLimitJoin) {
+		return fmt.Errorf("cannot change participation for this crowdaction anymore")
```

Comment:
please use github.com/pkg/errors  to create new error (`fmt.Errorf` will not carry stacktrace)

Reply:
done